### DISCUSSION TITLE
Keycloak Operator Documentation

### DISCUSSION THREAD

**Discussion by ML-std** [keycloak/keycloak#21839] [Ideas]
Keycloak Operator has examples for use cases. However, its not enough. If we have documentation for Keycloak Operator something like this:
https://www.keycloak.org/server/all-config
it can be better for use. for example, I search for all configurations when I want to deploy a custom Keycloak image, however this link is not enough, and I think there are more configurations more than this link below:
https://www.keycloak.org/operator/advanced-configuration

Also we may consider creating a Helm chart using Keycloak object.

**Comment by vmuzikar** [comment]
You can use the [`additionalOptions` field](https://www.keycloak.org/operator/advanced-configuration#_additional_options) for specifying options that are not available as first class citizens in the Keycloak CR.

We currently do not plan introducing Helm charts.

**Comment by ML-std** [comment]
Thanks, the problem I encountered was mostly about the documentation. I had to find from CRDS to understand how to autoscale and I learned that I should define instance field in that object. I solved my problems at this point, but I think it would be better to have all the available configs in one yaml, or even in one page like this: https://www.keycloak.org/operator/advanced-configuration. I see you have added almost all of them, but I'm not sure If there are any missing fields (like instance).

**Reply by vmuzikar** [reply]
You can directly autoscale the CRD using an HPA.

You can see all available options (incl. descriptions) in the CRD:
```
kubectl describe crd keycloaks.k8s.keycloak.org
```
You may notice that the structure and options in the CRD are closely aligned with the [server options](https://www.keycloak.org/server/all-config).

Feel free to open an issue for any docs improvements.

**Comment by ML-std** [comment]
Yes, we may consider creating an issue. For example, I cannot add my custom spi-login-provider as custom-freemarker to my custom keycloak image. I'm adding this line:

spec:
  additionalOptions:
...
      spi-login-provider: custom-freemaker
...

my custom-freemarker is not implemented in keycloak-operator. And Im gettiing NullPointer exception. Without the keycloak operator, it works fine, but I want it to be scalable.

What am i missing? should i create an issue for it?

**Reply by vmuzikar** [reply]
Can you please share the options and command that works when executed locally but doesn't with the Operator?

**Reply by ML-std** [reply]
As far as I understand, keycloak-operator works with --optimized flag. When I use operator, these doesnt work:

database:
  username: username
  password: password
  host: my-host.com
  database: "postgres"
  port: "5432"
  schema: "keycloak"

http:
  httpEnabled: false
  tlsSecret: dev-cert-prod

hostname:
  strictBackchannel: "true"

additionalOptions:
 spi-login-provider: custom-freemaker
 proxy: reencrypt
 https-certificate-key-file: /opt/keycloak/conf/tls.key
 https-certificate-file: /opt/keycloak/conf/tls.crt
 health-enabled: true
 hostname-url: https://my-host-url.com
 hostname-admin-url: https://my-host-url.com
 hostname-strict-backchannel: true
 
 image.name: my-custom-keycloak-image:latest
 
 
these are values for helm chart. there are rendered correctly on the pod's environment. however, when using operator, keycloak starts, it doesnt implement my custom-freemaker. Pod gives healthcheck status up, but doesnt get my provider as a result.

I can provide you pod's yaml if you need.

**Reply by vmuzikar** [reply]
If you are using a custom Keycloak image, the Operator assumes it's already optimized/augmented. If you need to specify any build-time options, those need to be defined directly in the container image. Please see the relevant [docs](https://www.keycloak.org/operator/customizing-keycloak).

**Reply by ML-std** [reply]
thanks for reply. I've read that and ı also tried to build with this command below (from dockerfile) when I try, even though keycloak understands the jar, it fails for not finding the provider. this is the Dockerfile:

FROM quayio-proxy/keycloak/keycloak:22.0.1 as builder

ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

ENV KC_DB=postgres
COPY --chown=keycloak:root ./keycloak-provider/build/libs /opt/keycloak/providers
RUN ls /opt/keycloak/providers
RUN /opt/keycloak/bin/kc.sh build --spi-login-provider=custom-freemaker --spi-login-provider-custom-freemaker-enabled=true

**Reply by ML-std** [reply]
I have error something like this:
Build failure: Build failed due to errors
	[error]: Build step org.keycloak.quarkus.deployment.KeycloakProcessor#configureKeycloakSessionFactory threw an exception: java.lang.RuntimeException: Failed to find provider custom-freemaker for login

**Reply by vmuzikar** [reply]
@ML-std I see. It's a bit hard to tell what went wrong there without knowing the contents of your `./keycloak-provider/build/libs` directory. Would you be able to provide a reproducer?

**Reply by ML-std** [reply]
Thats a build jar from gradle. I'm sorry, I'm new to Keycloak, I'm mostly working on operating things. I just double checked this location that you mentioned, and there is a jar that is working.  And I checked the build logs as well, keycloak finds the jar and intterracts with it. at this location, there is only 1 jar and that is used for custom login. 

Maybe its not related, but we were using 20.0 version of keycloak earlier, and provider is written for it. We updated to 22, and in local, it works with not --optimized flag. Are there any change for finding the providerName from 20.0 to 22.01 at keycloak that we may not forget to update?

Again, sorry for this spesific case, but maybe we may find a good feature/bug from this problem.

**Reply by vmuzikar** [reply]
Seems to me the issue is with copying the provider JAR to the container. Try using `--chown=keycloak:keycloak`.

**Reply by ML-std** [reply]
done that, didn't work. However, I'm already checking with RUN ls command 
Step 7/12 : RUN ls /opt/keycloak/providers
 ---> Running in 188622e17114
keycloak-provider.jar
README.md

and i see keycloak find the jar from this log below:

2023-07-26 16:16:47,461 DEBUG [io.quarkus.deployment.index.ApplicationArchiveBuildStep] (build-36) Indexing dependency: /opt/keycloak/lib/../providers/keycloak-provider.jar

but it fails with exception above.

**Reply by ML-std** [reply]
Im wondering if this approach changed in 22.0.1. Because I'm out of solutions. This is from the provider.
```

package login;

import org.keycloak.Config;
import org.keycloak.forms.login.LoginFormsProvider;
import org.keycloak.forms.login.LoginFormsProviderFactory;
import org.keycloak.models.KeycloakSession;
import org.keycloak.models.KeycloakSessionFactory;

/**
 * Factory class for Custom Freemarker Login Forms Provider.
 * Basic configs for that provider are implemented here.
 */
public class CustomFreeMarkerLoginFormsProviderFactory implements LoginFormsProviderFactory {

    /**
     * Returns the implementation of the custom provider.
     *
     * @param session KeycloakSession
     * @return LoginFormsProvider
     */
    @Override
    public LoginFormsProvider create(KeycloakSession session) {
        return new CustomFreeMarkerLoginFormsProvider(session);
    }

    @Override
    public void init(Config.Scope config) {
        // Comes from the Keycloak interface. No operation for this provider.
    }

    @Override
    public void postInit(KeycloakSessionFactory factory) {
        // Comes from the Keycloak interface. No operation for this provider.
    }

    @Override
    public void close() {
        // Comes from the Keycloak interface. No operation for this provider.
    }

    @Override
    public String getId() {
        return "custom-freemarker";
    }
}

```

**Comment by ML-std** [comment]
@vmuzikar For your information, doing this rather than this above in dockerfile solved our problem. We may consider updating doc for this, or solving this issue:

```
FROM quayio-proxy/keycloak/keycloak:22.0.1 as builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

WORKDIR /opt/keycloak
# Configure a database vendor
ENV KC_DB=postgres
ENV KC_SPI_LOGIN_PROVIDER=custom-freemarker

COPY --chown=keycloak:keycloak ./keycloak-provider/build/libs /opt/keycloak/providers
RUN /opt/keycloak/bin/kc.sh build --spi-login-provider-custom-freemaker-enabled=true
```

Giving as environment variable rather than build directly solved our problem.
